### DISCUSSION TITLE
problems using preview mode

### DISCUSSION THREAD

**Discussion by dohomi** [vercel/next.js#11315] [Help]
Hello,

I started to convert my SSR project to SSG and try to implement the preview mode. Does the preview mode intend to work while in development mode as well? Because I can't see the cookie being set. I am using a very basic approach and deployed via Vercel:
```js
// pages/api/preview
export default (req: NextApiRequest, res: NextApiResponse) => {
  if (!req.query.slug) {
    return res.status(401).json({ message: 'Invalid slug' })
  }
  res.setPreviewData({})
  res.writeHead(307, { Location: req.query.slug})
  res.end()
}
```
If I log the value of `context.preview` inside of my getStaticProps it is undefined and thus the preview does not work.
```
[ event ] build page: /api/preview
[ wait ]  compiling ...
[ ready ] compiled successfully - ready on http://localhost:3000
inside preview home // => this is home inside of api/preview
inside index
preview { index: [ 'api', 'home' ] } undefined undefined // preview slug is "api/home" and preview as well as preview data is undefined
{ status: 404, url: '/cdn/stories/api/home' }

```
Is there any way how to debug this behaviour? Thanks!

**Comment by dohomi** [comment]
I found the issue being a missing leading slash. Beside that it seems that only in production builds the preview mode is being triggered what is fine for me. The SSG feature is simply amazing!

**Reply by timneutkens** [reply]
> Beside that it seems that only in production builds the preview mode is being triggered what is fine for me.

Preview mode works in development too

**Reply by dohomi** [reply]
@timneutkens in development mode preview and previewData remains `undefined`, thus deployed on the server it is set

**Reply by timneutkens** [reply]
Maybe the cookies are not being set based on the domain that you're using. It definitely works in development if the cookies are set correctly.

**Reply by dohomi** [reply]
I only followed the preview mode guide and haven't done anything specific. I use following `pages/api/preview.ts`
```js
import { NextApiRequest, NextApiResponse } from 'next'

export default (req: NextApiRequest, res: NextApiResponse) => {
  let currentSlug = req.query.slug
  if (!currentSlug || typeof currentSlug !== 'string') {
    return res.status(401).json({ message: 'Invalid slug' })
  }
  currentSlug = currentSlug.startsWith('/') ? currentSlug : `/${currentSlug}`
  res.setPreviewData({
    query: req.query
  })
  res.writeHead(307, { Location: currentSlug })
  res.end()
}
```
I don't manipulate cookies or working with cookies. the `next dev` runs on localhost:3000

**Comment by dohomi** [comment]
@timneutkens I think I found the issue: I am using Storyblok headless CMS and their visual composer is rendering my page inside an iframe. So in the backend of Storyblok the cookies are not set, if I use it outside of Storyblok then it works.
But thats exactly what I need: a solution that my visual composer works. Is there an alternative way (maybe with url params) to get the previewData to the `getStaticProps`?

**Reply by ChrisDalley** [reply]
I'm having the same issue with iFrame'd previews in Storyblok causing cookies not be set. From what I can see looking into the implementation, there is no easy work around without using a custom server (though I may be missing something obvious).

**Comment by dohomi** [comment]
@timneutkens I created an issue because I couldn't find a solution for this problem: https://github.com/zeit/next.js/issues/11356

**Reply by peps** [reply]
I came here because I had the same issue. Turns out I was using Brave browser and it was blocking the third party cookies and localstorage access. If you're using Brave make sure to turn off Brave Shield on storyblok!

**Comment by gitabhisingh** [comment]
After curl I found out that cookies are being set but in **SameSite=None** mode. Is that the issue?
I am using Next JS v9.5.2

```
 HTTP/1.1 200 OK
HTTP/1.1 200 OK
< Set-Cookie: __prerender_bypass=e00e4ba3b660386df18eea2687f9b75; Path=/; HttpOnly; Secure; SameSite=None
Set-Cookie: __prerender_bypass=e00e4ba3b660386ef18eea2687f9b75; Path=/; HttpOnly; Secure; SameSite=None
< Set-Cookie: __next_preview_data=eyJhbGciOiJIUzI1NiIsInR5cI6IkpXVCJ9.eyJkYXRhIjoiYTU0NWQyN2RkZTM1ZTBiNjhiNzE3MTBkOWMxZjBmMGVlZDU3ZjEwMGQ5ZmExZTIxYmRjMDliMDY4Yjc5ODA1MWNlYzdhYWU1OWY5MzY1MTI5Y2RiMTNlZTA2OTM3M2ZhYjVkYjQ3NjYxYjdlMTkzMThhNmM4N2MxY2MzZTI3NDVhND4YjI0ZjllODgxN2I2OWYxODdhMTFhYjE0YzUyNzE0ZDQ0NDI4OTI1MWVhOTg1MDdkNWZlMDZkNThiNmUzMzI2MSIsImlhdCI6MTYwMzQ2NzA2Nn0.V4IpUW7V_goPgiJ5mV6-axAKQcMFxqNXwUQqVp6nfA; Path=/; HttpOnly; Secure; SameSite=None
Set-Cookie: __next_preview_data=eyJhbGciOiJIUzI1NisInR5cCI6IkpXVCJ9.eyJkYXRhIjoiYTU0NWQyN2RkZTM1ZTBiNjhiNzE3MTBkOWMxZjBmMGVlZDU3ZjEwMGQ5ZmExZTIxYmRjMDliMDY4Yjc5ODA1MWNlYzdhYWU1OWY5MzY1MTI5Y2RiMTNlZTA2OTM3M2ZhYjVkYjQ3NjYxYjdlMTkzMThhNmM4N2MxY2MzZTI3NDVhNDk4YjI0ZjllODgxN2I2OWYxODdhMTFhYjE0YzUyNzE0ZDQ0NDI4OTI1MWVhOTg1MDdkNWZlMDZkNThiNmUzMzI2MSIsmlhdCI6MTYwMzQ2NzA2Nn0.V4IpUW7V_goPgiJ5mV6-axAKQcMFxqNXwWQqVp6nfA; Path=/; HttpOnly; Secure; SameSite=None
< Date: Fri, 23 Oct 2020 15:31:06 GMT
Date: Fri, 23 Oct 2020 15:31:06 GMT
< Connection: keep-alive
Connection: keep-alive
< Transfer-Encoding: chunked
Transfer-Encoding: chunked

< 
<!DOCTYPE html><html><head><meta http-equiv="Refresh" content="0; url=/posts/everything-you-want-to-know" />
    <script>window.location.href = '/posts/everything-you-want-to-know'</script>
* Connection #0 to host localhost left intact
    </head>* Closing connection 0
```

**Comment by ayikfour** [comment]
Any soloution for this one? I'm stuck with this problem too. Preview value get undefined in Production/after build.

**Comment by AlexandraKlein** [comment]
Also finding query params undefined in production.

```javascript
export default async function preview(req, res) {
  const { slug, id } = req.query;
  console.log({ slug, id }); // undefined in production. Works as expected in development
}
```

**Comment by frankie567** [comment]
I also had this issue in a local development environment.

The thing is the cookies were correctly set in the response but they were in `Secure` mode ; which require an HTTPS/SSL connection, otherwise the browser ignores them. Obviously, in local development, I didn't have HTTPS setup.

I found out the issue by looking at NextJS code of the `setPreviewData` function:

https://github.com/vercel/next.js/blob/7076758c8d0534a7745bc7f6142d0ac85d9e5a7b/packages/next/next-server/server/api-utils.ts#L430

So the cookie is set in `Secure` mode if  the environment variable `NODE_ENV` is not equal to `development`. In my case, `NODE_ENV` was not set correctly. Setting it correctly with the value `development` immediately solved the issue.

**Reply by leerob** [reply]
Updated docs based on this discussion: https://github.com/vercel/next.js/pull/22906
